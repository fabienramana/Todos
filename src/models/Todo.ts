export default interface Todo{
    content: string,
    completed: boolean
}