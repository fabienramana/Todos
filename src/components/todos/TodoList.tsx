import { useRef } from "react";
import Todo from "../../models/Todo"
import TodoItem from './TodoItem'

type AppProps = {
    todoArray: Todo[],
    removeTodo: (index:number) => void,
    changeStatusOfTodo: (todoStatus: boolean, index: number) => void,
    changeStatusOfAllTodos: (status: boolean) => void, 
    filter: string
}

export default function TodoList({ todoArray, removeTodo, changeStatusOfTodo, changeStatusOfAllTodos, filter }: AppProps): JSX.Element{

    const inputRef = useRef(document.createElement("input"));

    function updateInputValue(){
        inputRef.current.checked = !inputRef.current.checked
        changeStatusOfAllTodos(inputRef.current.checked)
    }

    return(
        <section className="main">
				<input id="toggle-all" className="toggle-all" type="checkbox" onClick={(e) => changeStatusOfAllTodos(e.currentTarget.checked)}/>
				<label onClick={() => updateInputValue()}>Mark all as complete</label>
				<ul className="todo-list">
					{todoArray.filter((todo)=>{
                        if(filter === "active" && todo.completed === false){
                            return true;
                        }
                        else if(filter === "completed" && todo.completed === true){
                            return true;
                        }
                        else if(filter === "all"){
                            return true;
                        }
                        return false;
                    }).map((todo, index) => <TodoItem key={index}
                                                     value={index}
                                                     todo={todo}
                                                     removeTodo={removeTodo}
                                                     changeStatusOfTodo={changeStatusOfTodo}/>)}
				</ul>
			</section>
    )
}