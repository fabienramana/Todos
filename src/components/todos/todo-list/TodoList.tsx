import Todo from "../../../models/Todo"
import TodoItem from './TodoItem'
import classNames from 'classnames';
import { useState } from "react";

type TodoListProps = {
    todos: Todo[],
    removeTodo: (id: number) => void,
    changeTodo: (todo: Partial<Todo>, id: number) => void
}

export default function TodoList({ todos, removeTodo, changeTodo }: TodoListProps){
    const [editModeIndex, setEditMode] = useState<number>(0);

    const resetEditMode = () => {
        setEditMode(0)
    }

    return(		
				<ul className="todo-list">
					{todos.map((todo) => {
                        const myComponentClasses = classNames({
                            "completed": todo.completed,
                            "editing": editModeIndex === todo.id
                        })
                        return(<li key={todo.id} className={myComponentClasses}>
                                    <TodoItem 
                                            todoTitle={todo.title}
                                            todoStatus={todo.completed}
                                            removeTodo={() => removeTodo(todo.id)}
                                            changeTodo={(t: Partial<Todo>, id=todo.id) => changeTodo(t, id)}
                                            setEditMode={() => setEditMode(todo.id)}
                                            resetEditMode={resetEditMode} />
                               </li>)
                    })}
				</ul>
    )
}