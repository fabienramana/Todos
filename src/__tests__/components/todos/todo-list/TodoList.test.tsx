import { render, screen } from "@testing-library/react"
import TodoList from "../../../../components/todos/todo-list/TodoList"
import Todo from "../../../../models/Todo"



describe('TodoList component', () => {

    const listOfTodos: Array<Todo> = [
      {
          id:1,
          title:"Wake up early",
          completed: false
      },
      {
          id:2,
          title:"Sleep early",
          completed: false
      },
      {
          id:3,
          title:"Do the chores",
          completed: true
      } 
  
  ]
    
    test('should display all todos content', () => {
  
      const removeTodo = () => {}
      const changeTodo = () => {}
  
      render(<TodoList todos={listOfTodos} 
        removeTodo={removeTodo}
        changeTodo={changeTodo}/>)
      
      listOfTodos.forEach(todo => {
        expect(screen.getByText(todo.title)).toBeInTheDocument();
      })
    })

    test('should match snapshot', () => {
  
      const removeTodo = () => {}
      const changeTodo = () => {}
  
      const {asFragment} = render(<TodoList todos={listOfTodos} 
        removeTodo={removeTodo}
        changeTodo={changeTodo}/>)
  
      expect(asFragment()).toMatchSnapshot();
    })

  })