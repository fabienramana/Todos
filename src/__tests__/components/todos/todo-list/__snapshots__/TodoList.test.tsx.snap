// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TodoList component should match snapshot 1`] = `
<DocumentFragment>
  <ul
    class="todo-list"
  >
    <li
      class=""
    >
      <div>
        <div
          class="view"
        >
          <input
            aria-label="changeStatus"
            class="toggle"
            type="checkbox"
          />
          <label
            aria-label="title"
          >
            Wake up early
          </label>
          <button
            aria-label="delete"
            class="destroy"
          />
        </div>
        <input
          class="edit"
          value="Wake up early"
        />
      </div>
    </li>
    <li
      class=""
    >
      <div>
        <div
          class="view"
        >
          <input
            aria-label="changeStatus"
            class="toggle"
            type="checkbox"
          />
          <label
            aria-label="title"
          >
            Sleep early
          </label>
          <button
            aria-label="delete"
            class="destroy"
          />
        </div>
        <input
          class="edit"
          value="Sleep early"
        />
      </div>
    </li>
    <li
      class="completed"
    >
      <div>
        <div
          class="view"
        >
          <input
            aria-label="changeStatus"
            checked=""
            class="toggle"
            type="checkbox"
          />
          <label
            aria-label="title"
          >
            Do the chores
          </label>
          <button
            aria-label="delete"
            class="destroy"
          />
        </div>
        <input
          class="edit"
          value="Do the chores"
        />
      </div>
    </li>
  </ul>
</DocumentFragment>
`;
